### ISSUE TITLE
Python example configuration

### ISSUE THREAD

**Issue by 42-dd** [cbdevnet/midimonster#83]
Hi there.
I'm pretty new to python programming but am willing too learn. For now I seem to be struggling to set up the python backend in midi monster correctly. What I want to do:
- receive ArtNet or OSC
- convert the values via a python script (example: ArtNet Pan Coarse & Fine combined to Pan)
- put out OSC and VISCA

Is there any example config that could nudge me in the right direction?

Thanks for your great work.

**Comment by cbdevnet** [comment]
Hi @dd253, and welcome to the MIDIMonster :)

You're right, there's currently no example config for python - I'll fix that soon!

Python code is frequently organized into "modules", which are files containing functions that deal with a common theme. In the MIDIMonster, `python` instances can load one or more python modules to execute.

When events are then mapped onto the instance, functions with the corresponding name (or the default handler if one is set) are executed with the value as the argument, and can then do anything they want with it, such as calculate some stuff, output them again, store the somewhere, etc..

A very basic python module for MIDIMonster could look like this:

```python
import midimonster

def channel1(value):
    print("Value is %s\n" % (value,))
    midimonster.output("out1", 1 - value)
```
Saving that file as "mymodule.py" in the same folder as the following demo configuration:

```
[python py1]
module = mymodule

[backend artnet]
bind = 0.0.0.0

[artnet a1]
dest = 255.255.255.255
universe = 1

[map]
a1.1 > py1.mymodule.channel1
```

and running the configuration will print the current value on the console when it changes, and the channel `py1.out1` will have the value `1 - input`

**Comment by dd253** [comment]
Thanks for pointing this out and for your quick response. I'll give it a try and get back to you.

**Comment by dd253** [comment]
Hi. Thanks for your help. I got the example running, but I'm not able to get the python output py1.out1 mapped to OSC.

Here's one config I tried:


[python py1]
module = mymodule


[backend osc]
detect = off

[osc a1]
bind = * 8000
dest = 192.168.5.4 8001
/2* = f 0.0 1.0
;/3* = f 0.0 1.0


[map]
a1./xy:0 > py1.mymodule.channel1
py1.out1 > a1./2
;a1./xy:0 > a1./3

Any ideas what I'm doing wrong?

**Comment by phedders** [comment]
On Tue, 2021-01-19 at 04:47 -0800, dd253 wrote:
> Hi. Thanks for your help. I got the example running, but I'm not able
> to get the python output py1.out1 mapped to OSC.
> Here's one config I tried:
> [python py1]
> module = mymodule
> [backend osc]
> detect = off
> [osc a1]
> bind = * 8000
> dest = 192.168.5.4 8001
> /2* = f 0.0 1.0
> ;/3* = f 0.0 1.0
> [map]
> a1./xy:0 > py1.mymodule.channel1
> py1.out1 > a1./2
> ;a1./xy:0 > a1./3
> Any ideas what I'm doing wrong?
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

[python py1]
> module = mymodule

Is that the full path?

[python py1]
> module = mymodule.py

seems more likely to me

**Comment by phedders** [comment]
On Tue, 2021-01-19 at 04:47 -0800, dd253 wrote:
> Hi. Thanks for your help. I got the example running, but I'm not able
> to get the python output py1.out1 mapped to OSC.
> Here's one config I tried:
> [python py1]
> module = mymodule
> [backend osc]
> detect = off
> [osc a1]
> bind = * 8000
> dest = 192.168.5.4 8001
> /2* = f 0.0 1.0
> ;/3* = f 0.0 1.0
> [map]
> a1./xy:0 > py1.mymodule.channel1
> py1.out1 > a1./2
> ;a1./xy:0 > a1./3
> Any ideas what I'm doing wrong?
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.


Also I havent tried python modules properly - but it might require

	script = <path>

like the lua backend.

The lua backend is very lightweight and fast by the way... for simple
translations I would suggest it might be more appropriate.

**Comment by dd253** [comment]
That doesn't seem to make any difference. It loads the correct module anyway....

**Comment by cbdevnet** [comment]
It seems that there is currently a problem with the `python` backend, as things that once worked in my tests seem to be broken by an update within Python. Normally, that API should have been stable within my Debian system, but it seems something has changed anyway.

I'll investigate further and report back.

**Comment by cbdevnet** [comment]
The problems in reproducing the test configuration should have been resolved with 011be343cd1427a1be68f9a0da38401b89de0fec, I added an example configuration using both lua and python in a92c6fb9f20ae26335b2ea3bd4539dd6e4aec256. The backend documentation has also been updated to be a bit clearer.
